ref_builder/index.py

```diff
@@ -20,76 +28,373 @@ class EventIndexItem:
+        updated.
+
+        Only sequences that have changed will be updated.
 
-        :param otu_id: A Virtool OTU Id
-        :return: An OTUEventCache object or None
         """
-        path = self.path / f"{otu_id}.json"
+        sequence_ids = {
+            str(seq.id) for isolate in otu.isolates for seq in isolate.sequences
+        }
+
+        otu_dict = otu.model_dump()
+
+        placeholders = ",".join("?" for _ in sequence_ids)
 
-        try:
-            with open(path, "rb") as f:
-                return EventIndexItem(
-                    **{**orjson.loads(f.read()), "otu_id": otu_id},
+        # Delete any sequences that are no longer in the OTU.
+        self.con.execute(
+            f"""
+            DELETE FROM sequences WHERE otu_id = ? AND NOT id IN ({ placeholders });
+            """,
+            (str(otu.id), *sequence_ids),
+        )
+
+        batch = []
+
+        # Insert or update the sequences.
+        for isolate in otu_dict["isolates"]:
+            for sequence in isolate["sequences"]:
+                crc = _calculate_crc32(sequence["sequence"])
+                batch.append(
+                    (
+                        str(sequence["id"]),
+                        crc,
+                        str(otu.id),
+                        sequence["sequence"],
+                    ),
                 )
-        except FileNotFoundError:
-            return None
+
+        # Perform batch insert
+        self.con.executemany(
+            """
+            INSERT OR REPLACE INTO sequences (id, crc, otu_id, sequence)
+            VALUES (?, ?, ?, ?)
+            """,
+            batch,
+        )
+
+        # Update only if CRC has changed
+        self.con.executemany(
+            """
+            UPDATE sequences
+            SET crc = ?, otu_id = ?, sequence = ?
+            WHERE id = ? AND crc != ?
+            """,
+            [(crc, otu_id, seq, seq_id, crc) for seq_id, crc, otu_id, seq in batch],
+        )
+
+        self.con.execute(
+            """
+            INSERT OR REPLACE INTO otus (id, acronym, at_event, legacy_id, name, otu, taxid)
+            VALUES (?, ?,?, ?, ?, ?, ?)
+            """,
+            (
+                str(otu.id),
+                otu.acronym,
+                at_event,
+                otu.legacy_id,
+                otu.name,
+                orjson.dumps(otu_dict, default=_default_json),
+                otu.taxid,
+            ),
+        )
+
+        self.con.commit()
+
+
+def _calculate_crc32(sequence: str) -> str:
+    """Calculate the CRC32 checksum of a sequence.
+
+    :param sequence: the sequence to calculate the checksum of.
+    :return: the CRC32 checksum as a hexadecimal string.
+
+    """
+    crc = binascii.crc32(sequence.encode("utf-8"))
+
+    # Convert CRC as a hexadecimal string.
+    return hex(crc & 0xFFFFFFFF)[2:].zfill(8)
+
+
+def _default_json(obj: Any) -> list:
+    """Cast sets to lists.
```

Comment:
A minor note: [`.model_dump(mode="json")`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_dump) automatically handles this cast, as well as the UUID cast. It can also be handled by a field serializer.

Reply:
Sweet. I've made the change.